src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -56,211 +88,802 @@ protected Marshaller()
+            if (parameterType.IsByRef)
+            {
+                // Passing as [Out] by ref is valid
+                marshaller.Out = pInvokeParameterdata.Out;
+            }
+            else
+            {
+                // Passing as [Out] is valid only if it is not ValueType nor string
+                if (!parameterType.IsValueType && !parameterType.IsString)
+                    marshaller.Out= pInvokeParameterdata.Out;
+            }
 
+            if (!marshaller.In && !marshaller.Out)
+            {
+                //
+                // Rules for in/out
+                // 1. ByRef args: [in]/[out] implied by default
+                // 2. StringBuilder: [in, out] by default
+                // 3. non-ByRef args: [In] is implied if no [In]/[Out] is specified
+                //
+                if (parameterType.IsByRef)
+                {
+                    marshaller.In = true;
+                    marshaller.Out = true;
+                }
+                else if (pInvokeMethodData.IsStringBuilder(parameterType))
+                {
+                    marshaller.In = true;
+                    marshaller.Out = true;
+                }
+                else
+                {
+                    marshaller.In = true;
+                }
+            }
             return marshaller;
         }
 
         private static Marshaller CreateMarshallerInternal(MarshallerKind kind)
         {
             switch (kind)
             {
+                case MarshallerKind.Enum:
                 case MarshallerKind.BlittableValue:
+                case MarshallerKind.BlittableStruct:
+                case MarshallerKind.UnicodeChar:
                     return new BlittableValueMarshaller();
+                case MarshallerKind.Array:
                 case MarshallerKind.BlittableArray:
                     return new BlittableArrayMarshaller();
-                case MarshallerKind.BlittableByRef:
-                    return new BlittableByRefMarshaller();
                 case MarshallerKind.Bool:
                     return new BooleanMarshaller();
-                case MarshallerKind.String:
-                    return new StringMarshaller();
+                case MarshallerKind.AnsiString:
+                    return new AnsiStringMarshaller();
+                case MarshallerKind.UnicodeString:
+                    return new UnicodeStringMarshaller();
                 case MarshallerKind.SafeHandle:
                     return new SafeHandleMarshaller();
-                case MarshallerKind.StringBuilder:
-                    return new StringBuilderMarshaller();
+                case MarshallerKind.UnicodeStringBuilder:
+                    return new UnicodeStringBuilderMarshaller();
                 case MarshallerKind.VoidReturn:
                     return new VoidReturnMarshaller();
                  default:
-                    throw new NotSupportedException();
+                    return new InvalidMarshaller();
```

Comment:
This was done to make sure the IsStubRequired code path doesn't throw, only EmitIL code path throws. Now that both are done in the same call it doesn't add any value. I will change it throw here.